test/topology_custom/test_writes_to_previous_cdc_generations.py

```diff
@@ -0,0 +1,140 @@
+        # Writes to all three generations should succeed.
+        await do_write(ts + 1)
+        await do_write(ts)
+        # A write with a timestamp smaller than the oldest generation's one should fail.
+        if ts == min(gen_timestamps):
+            with pytest.raises(NoHostAvailable):
+                await do_write(ts - 1)
+        else:
+            await do_write(ts - 1)
+
+
+@pytest.mark.asyncio
+@skip_mode('debug', 'test requires nodes to be started quickly')
+async def test_writes_to_old_previous_cdc_generation(request, manager: ManagerClient):
+    """
+    Test that writes to a previous CDC generation succeed if the write's timestamp is greater than
+    (now - generation_leeway) even if the generation's timestamp is not greater than (now - generation_leeway).
+
+    The test starts 2 nodes, and thus, it creates 2 generations. Let's denote their timestamps by ts1 and ts2, where
+    ts1 is the timestamp of the first (and oldest) generation. We create a scenario where
+    ts1 < now - generation_leeway < ts2 < now.
+    We check that writes with a timestamp from the interval (now - generation_leeway, ts2) succeed.
+
+    ts2 might already be much smaller than now after we finish adding the second node. Then, a write with a timestamp
+    smaller than ts2 could unexpectedly fail because we reject writes to previous generations with timestamps not
+    greater than (now - generation_leeway). To make this situation less likely, we increase ring-delay to 5 s, which
+    increases ts2 by 15 s (see cdc::new_generation_timestamp). If this situation still happens, the test doesn't execute
+    writes to avoid flakiness. We don't want to increase ring-delay more as it would slow down the test even more. To
+    sum up, the test shouldn't be flaky, but there is no guarantee that it will always test what it's supposed to test.
+    We also ignore it in debug mode due to the reasons above.
+    """
+    logger.info("Bootstrapping nodes")
+    servers = await manager.servers_add(2, cmdline=['--ring-delay', '5000'])
+
+    cql = manager.get_cql()
+
+    logger.info("Waiting for driver")
+    await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
+
+    gen_timestamps = await wait_for_publishing_generations(cql, servers)
+
+    logger.info("Ceating a test table")
+    await cql.run_async("CREATE KEYSPACE test WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1}")
+    await cql.run_async("CREATE TABLE test.test (pk int PRIMARY KEY, c int) WITH cdc = {'enabled': true}")
+
+    ts2 = max(gen_timestamps)
+    time_diff = (ts2 / (1000 * 1000)) - time.time()
+    if time_diff > 0:
+        logger.info("Waiting for the second generation to become active")
+        await asyncio.sleep(time_diff)
+
+        # Writes with timestamps from the interval (now - generation_leeway, ts2) should succeed.
+        await cql.run_async(f"INSERT INTO test.test (pk, c) VALUES (0, 0) USING TIMESTAMP {ts2 - 1}")
```

Comment:
IIUC this could still be flaky.

I think a non-flaky way would be to catch the error, see if it's the error that should be thrown if `now - generation_leeway >= ts2`, verify that indeed `now - generation_leeway >= ts2`, and if so, skip the check. But if it turns out that `now - generation_leeway < ts2` and the error was still thrown, it's a bug.

So for example if the test hangs for a long time before this `run_async`, we will skip the check instead of failing.

Reply:
> IIUC this could still be flaky.

Yes, if handling a write can take more than 5 s, then it can still be flaky. I don't know if this is something we have to consider. You proposed that we should probably disallow all writes with `T < now - generation_leeway` in https://github.com/scylladb/scylladb/pull/17134#issuecomment-1923927162. If we do it, won't all writes with CDC become flaky?

> I think a non-flaky way would be to catch the error, see if it's the error that should be thrown if `now - generation_leeway >= ts2`, verify that indeed `now - generation_leeway >= ts2`, and if so, skip the check. But if it turns out that `now - generation_leeway < ts2` and the error was still thrown, it's a bug.

The problem is this verification part. It requires extracting `now` from the error message. And this `now` is formatted as date. I don't know how easy it is to implement. If doing something like this is the only option here, then I'll think about it.

Reply:
I can also get `now` directly in the test after catching the error and use it for verification. The test won't be flaky but it also won't detect all incorrectly thrown errors.

Reply:
I went for the latter option.